### PR TITLE
Handle EINTR and EAGAIN when we `poll`

### PR DESCRIPTION
**Pull request by FranzBusch**
# Motivation
We are currently making calls to `poll` which could get interrupted or the resource could be temporarily not available. We should handle these cases.

# Modification
Wrap the calls to `poll` in a `wile` loop and try again if we get either `EINTR` or `EAGAIN`.

# Result
We are now handling signals while we are polling correctly.

### CODE REVIEW

File: lib/Basic/Subprocess.cpp [hunk 1]
```diff
@@ -1056,11 +1056,20 @@ void llbuild::basic::spawnProcess(
     assert(readfds[0].fd == outputPipeParentEnd.unsafeDescriptor());
     assert(readfds[1].fd == controlPipeParentEnd.unsafeDescriptor());
 
-    if (poll(readfds, nfds, -1) == -1) {
-      int err = errno;
-      delegate.processHadError(ctx, handle,
-          Twine("failed to poll (") + strerror(err) + ")");
-      break;
+    while (true) {
+      if (poll(readfds, nfds, -1) == -1) {
```
Comment (Lukasa) [inline lib/Basic/Subprocess.cpp:1060]: You can move this `if` condition into the `while` condition, and avoid writing the `else break` and `while true`.

**Comment by weissi** [comment]
Please test

**Comment by weissi** [comment]
test this please

**Comment by dmbryson** [comment]
@swift-ci please smoke test

File: lib/Basic/Subprocess.cpp [hunk 2]
```diff
@@ -1056,11 +1056,16 @@ void llbuild::basic::spawnProcess(
...
-      break;
+    while (poll(readfds, nfds, -1) == -1) {
+        int err = errno;
+
+        if (err == EAGAIN || err == EINTR) {
+          continue;
+        } else {
+          delegate.processHadError(ctx, handle,
+            Twine("failed to poll (") + strerror(err) + ")"); 
+          goto pollError;
```
Comment (dmbryson) [inline lib/Basic/Subprocess.cpp:1067]: Can we just return here?   Why the `goto` ?

Comment (FranzBusch) [reply lib/Basic/Subprocess.cpp:1067]: I guess we can. I just wanted to have the exact same behaviour as before where we returned to after the while loop

Comment (dmbryson) [reply lib/Basic/Subprocess.cpp:1067]: I think the more direct return is better in this case and matches the coding style more cleanly.

**Comment by dmbryson** [comment]
@swift-ci please smoke test

**Review by dmbryson — APPROVED** [review]

**Review by jakepetroules — APPROVED** [review]

**Pull request merged**